# K-近邻算法

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#K-临近算法概述" data-toc-modified-id="K-临近算法概述-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>K-临近算法概述</a></span></li><li><span><a href="#实施KNN分类算法" data-toc-modified-id="实施KNN分类算法-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>实施KNN分类算法</a></span></li></ul></div>

## K-临近算法概述

以电影类型分类为例，我们可以明确每部电影在风格上会同相同题材的电影相近。例如动作片之间有什么共同的特征，这个特征又和爱情片之间存在明显的差异呢？

我们可以使用电影中打斗次数和接吻次数，使用KNN算法自动划分电影题材。

> KNN算法优缺点：

+ 优点：精度高、对异常值不敏感、无输入数据假定。
+ 缺点：计算复杂度高、空间复杂度高。
+ 适用数据范围：数值型和标称型。

> 假设数据情况如下：

|电影名称|打斗镜头|接吻镜头|电影类型|
|:-|:-:|:-:|:-:|
|*California Man*|3|104|爱情片|
|*He's Not Really into Dudes*|2|100|爱情片|
|*Beautiful Woman*|1|81|爱情片|
|*Kevin Longblade*|101|10|动作片|
|*Robo Slayer 3000*|99|5|动作片|
|*Amped II*|98|2|动作片|
|*?*|18|90|*未知*|

In [2]:
import KNN
group, labels = KNN.createDataSet()
group

array([[1. , 1.1],
       [1. , 1. ],
       [0. , 0. ],
       [0. , 0.1]])

In [3]:
labels

['A', 'A', 'B', 'B']

group存储数据,labels存储标签。(1, 1.1)分类为A

![](../data/pic/四点简单示例.png)

## 实施KNN分类算法 

对未知属性的数据集中每个点依次执行以下操作：
1. 计算已知类别数据集中每个点与当前点之间的距离
2. 按照距离递增依次排序
3. 选取与当前点距离最小的$k$个点
4. 确定前$k$个点所在类别出现的频率
5. 返回频率最高的类别作为当前点的预测分类